src/main/java/net/neoforged/neoforge/transfer/ResourceHandlerUtil.java

```diff
@@ -0,0 +1,565 @@
+                if (fromResource.isEmpty() || !filter.test(fromResource)) continue;
+
+                // check how much can be extracted
+                int extracted;
+                try (Transaction simulatedExtractTransaction = TransactionManager.open(transaction)) {
+                    extracted = from.extract(index, fromResource, amount - totalMoved, simulatedExtractTransaction);
+                }
+
+                try (Transaction transferTransaction = TransactionManager.open(transaction)) {
+                    // check how much can be inserted
+                    int inserted = to.insert(fromResource, extracted, transferTransaction);
+
+                    // extract it, or rollback if the amounts don't match
+                    if (from.extract(index, fromResource, inserted, transferTransaction) == inserted) {
+                        totalMoved += inserted;
+                        transferTransaction.commit();
+                        lastMovedResource = fromResource;
+                    }
+                }
+
+                if (amount == totalMoved) {
+                    // early return if nothing can be moved anymore
+                    return stackFactory.create(lastMovedResource, totalMoved);
+                }
+            }
+            return stackFactory.create(totalMoved == 0 ? defaultResource : lastMovedResource, totalMoved);
+        } catch (Exception e) {
+            CrashReport report = CrashReport.forThrowable(e, "Moving resources between storages");
+            //noinspection DataFlowIssue
+            report.addCategory("Move details")
+                    .setDetail("Input", from::toString)
+                    .setDetail("Output", to::toString)
+                    .setDetail("Filter", filter::toString)
+                    .setDetail("Amount", amount)
+                    .setDetail("Transaction", transaction);
+            throw new ReportedException(report);
+        }
+    }
+
+    public static <T extends IResource> long getAmountAsLong(IResourceHandler<T> handler) {
+        long sum = 0L;
+        int size = handler.size();
+        for (int index = 0; index < size; index++) {
+            sum += handler.getAmountAsLong(index);
+            if (sum < 0) return Long.MAX_VALUE;
+        }
+        return sum;
+    }
+
+    public static <T extends IResource> long getCapacityAsLong(IResourceHandler<T> handler) {
+        long sum = 0L;
+        int size = handler.size();
+        for (int index = 0; index < size; index++) {
+            sum += handler.getCapacityAsLong(index, handler.getResource(index));
+            if (sum < 0) return Long.MAX_VALUE;
+        }
+        return sum;
+    }
+
+    /**
+     * @return {@code true} if the given resource is in the resource handler (though not necessarily interactable), {@code false} otherwise
+     */
+    public static <T extends IResource> boolean contains(IResourceHandler<T> handler, T resource) {
+        int size = handler.size();
```

Comment:
Given it short circuits the impl of this method could just be `return indexOf(handler, resource) != -1;`